Add benchmarks for SVE `fmla` instructions

dssgabriel opened the pull request:
This PR adds benchmarks for the throughput and latency of SVE FMA (`fmla`) instructions.

Sample results on NVIDIA Grace CPU (4x128-bit SVE):
```
SVE vector FMA per clk: 2.50
SVE vector FMA latency: 4.00 clocks
```

clamchowder commented:
Since SVE vector width isn't fixed across implementations, is there a way to provide throughput in operations per cycle?
Variable vector width is part of why I never got around to adding SVE support

dssgabriel commented:
I can add an intrinsic call that retrieves the number of active elements (of a given size/datatype) in an SVE register, which we can then use to correctly compute the ops/cycle throughout.

This could look like the following in assembly (for FP32):
```
sveops:
    cntw x0
    ret
```
Or just use the Arm intrinsic for it:
```c
const uint64_t sveops = svcntw();
```
See [the Arm docs](https://developer.arm.com/documentation/ddi0602/2023-09/SVE-Instructions/CNTB--CNTD--CNTH--CNTW--Set-scalar-to-multiple-of-predicate-constraint-element-count-) for more details on the instruction.

Also, I just realized that for wider SVE registers than NVIDIA Grace's 128-bit (e.g. AWS Graviton3 or Fujitsu A64FX), we will need to pass a bigger array of floats to the test function (8, 16 or more, instead of the current 4).

I will add all this to the PR tomorrow. Let me know if it sounds good to you 👍

clamchowder commented:
That sounds good. Can you also put it behind feature detection so it doesn't fault on Arm CPUs without SVE support? 

I did the same for x86 CPUs and AVX/AVX2/AVX-512, but I didn't do it for Arm because every aarch64 CPU supports ASIMD. However very few Arm CPUs right now support SVE.

dssgabriel commented:
I have added pre-processor guards for feature detection and changed the input array for the `fmla` SVE benchmarks.

I am unsure how you want to measure throughput other than instruction/cycle though. Whether SVE registers are 128- or 2048-bit wide shouldn't make any difference on what we're measuring, it's still a throughput in ops/cycle, right? :thinking: 
Unless what we really want is to measure the throughput in bytes/cycles that are processed by an instruction (e.g. `fmla` in this PR), I don't see how the code should change from the NEON/ASIMD test?

clamchowder commented:
Well it's easy to calculate computational throughput with fixed vector width instructions. For example 128-bit x86 SSE or Arm ASIMD instructions would process four 32-bit values at a time. Two of those instructions per cycle would mean 2*4=8 32-bit computations per cycle, or double that if you're using FMA and counting that as two. It was never a problem for Arm because they used fixed 128-bit vector lengths in NEON/ASIMD. For x86 it was never a problem because vector lengths were explicitly specified (`xmm`, `ymm`, `zmm`).

SVE makes it harder because `fmla zn.s` could operate on 4x 32-bit values (Grace). Or 8x 32-bit values (Graviton 3), or 16x 32-bit values (A64FX).

dssgabriel commented:
I was trying to say that if you are measuring instruction _throughput_, it does not matter how many items are being processed. We only measure how many independent instructions can be executed per cycle, regardless of how much data they operate on.
However, if we want to measure _bandwidth_, then the number of items being processed in a single instruction, of course, matters.

In the existing benchmarks for Arm ASIMD/NEON, you are only measuring throughput as **instruction/cycle**:
```
128-bit vector FMA per clk:  2.00
Scalar FMA per clk:          2.00
```
In this example, scalar FMA does 1/4 of the work of a 128-bit/NEON FMA but still has the same throughput.

In this case, the vector registers being fixed-width it is trivial to derive other metrics, such as **bytes/cycle**. And indeed, if we try to measure the latter when benchmarking SVE, it will be necessary to retrieve the µarch-specific register width to compute the metric correctly.

What I am trying to get at is if you want to measure the same thing as you already do for NEON, i.e. throughput as retired instructions/cycle, then this PR should not need further changes.   
If, however, you want to measure the bandwidth, i.e. the amount of data being processed as bytes/cycles, then I'll need to update the metric calculation to include the register width.

Could you please clarify which of the two we're trying to measure here?

dssgabriel commented:
Yes, this is definitely possible! :slightly_smiling_face:

My last commit now includes the vector register width for SVE FMA. I simply added a call to the intrinsic I mentioned in a previous comment (`svcntw`) when printing the results.

Samples results on two different SVE-enabled CPUs:
```sh
# AWS Graviton3 (1vCPU)
256-bit SVE vector FMA per clk: 1.28
256-bit SVE vector FMA latency: 4.00 clocks
```
```sh
# NVIDIA Grace Hopper Superchip
128-bit SVE vector FMA per clk: 2.49
128-bit SVE vector FMA latency: 4.01 clocks
```

I don't have access to an A64FX to test it out, but it should work the same and report "512-bit" :+1:
